packages/cli/src/lib/bundler/server.ts

```diff
@@ -21,29 +21,72 @@ import openBrowser from 'react-dev-utils/openBrowser';
 import { createConfig, resolveBaseUrl } from './config';
 import { ServeOptions } from './types';
 import { resolveBundlingPaths } from './paths';
+import { paths as libPaths } from '../../lib/paths';
+import { loadCliConfig } from '../config';
+import chalk from 'chalk';
+import { AppConfig } from '@backstage/config';
 
 export async function serveBundle(options: ServeOptions) {
-  const url = resolveBaseUrl(options.frontendConfig);
+  const { name } = await fs.readJson(libPaths.resolveTarget('package.json'));
+
+  let server: WebpackDevServer | undefined = undefined;
+  let latestFrontendAppConfigs: AppConfig[] = [];
+
+  const cliConfig = await loadCliConfig({
+    args: options.configPaths,
+    fromPackage: name,
+    withFilteredKeys: true,
+    watch(appConfigs) {
+      latestFrontendAppConfigs = appConfigs;
+      server?.invalidate();
+    },
+  });
+  latestFrontendAppConfigs = cliConfig.frontendAppConfigs;
+
+  const appBaseUrl = cliConfig.frontendConfig.getString('app.baseUrl');
+  const backendBaseUrl = cliConfig.frontendConfig.getString('backend.baseUrl');
+  if (appBaseUrl === backendBaseUrl) {
+    console.log(
+      chalk.yellow(
+        `⚠️   Conflict between app baseUrl and backend baseUrl:
+
+    app.baseUrl:     ${appBaseUrl}
+    backend.baseUrl: ${backendBaseUrl}
+
+    Must have unique hostname and/or ports.
+
+    This can be resolved by changing app.baseUrl and backend.baseUrl to point to their respective local development ports.
+`,
+      ),
+    );
+  }
```

Comment:
Only issue with this logic being moved out of `startFrontend.ts` and into `server.ts` is that validation logic now lives across 2 files, where previously it only done in `startFrontend.ts`, which left `server.ts` to handle the creation (and now restart) of the server.

Is there a way we can keep the `watch(appConfigs)` callback declared in this file but the `loadCliConfig` with its watch callback are still in `startFrontend.ts`? This would allow the validation to remaining in `startFrontend.ts`, and still only fetching the config once.